src/weather.cpp

```diff
@@ -66,6 +67,122 @@ static bool is_player_outside()
     return g->m.is_outside( point( g->u.posx(), g->u.posy() ) ) && g->get_levz() >= 0;
 }
 
+const weather_datum *weather::data( weather_type const type )
+{
+    const size_t i = static_cast<size_t>( type );
+    if( i < weather_datums.size() ) {
+        return { &weather_datums[i] };
+    }
+    debugmsg( "Invalid weather requested." );
+    return { &weather_datums[WEATHER_NULL] };
+}
+
+weather_type weather::get_bad_weather()
+{
+    weather_type bad_weather = WEATHER_NULL;
+    int current_weather = 0;
+    for( const weather_datum &weather : weather_datums ) {
+        current_weather++;
+        if( weather.precip == precip_class::HEAVY ) {
+            bad_weather = current_weather;
+        }
+    }
+    return bad_weather;
+}
+
+int weather::get_count()
+{
+    return weather_datums.size();
+}
+
+void weather::load_weather_type( const JsonObject &jo )
+{
+    weather_datum weather_data;
+    weather_data.id = jo.get_string( "id" );
+    weather_data.name = jo.get_string( "name" );
+    if( !assign( jo, "color", weather_data.color ) ) {
+        jo.throw_error( "missing mandatory member \"color\"" );
+    }
+    if( !assign( jo, "map_color", weather_data.map_color ) ) {
+        jo.throw_error( "missing mandatory member \"map_color\"" );
+    }
+    weather_data.glyph = jo.get_string( "glyph" )[0];
+    weather_data.ranged_penalty = jo.get_int( "ranged_penalty" );
+    weather_data.sight_penalty = jo.get_float( "sight_penalty" );
+    weather_data.light_modifier = jo.get_int( "light_modifier" );
+    weather_data.sound_attn = jo.get_int( "sound_attn" );
+    weather_data.dangerous = jo.get_bool( "dangerous" );
+    weather_data.precip = static_cast<precip_class>( jo.get_int( "precip" ) );
+    weather_data.rains = jo.get_bool( "rains" );
+    weather_data.acidic = jo.get_bool( "acidic" );
+    for( const JsonObject weather_effect : jo.get_array( "effects" ) ) {
+
+
+        std::pair<std::string, int> pair = std::make_pair( weather_effect.get_string( "name" ),
+                                           weather_effect.get_int( "intensity" ) );
+
+        std::map<std::string, weather_effect_fn> all_weather_effects = {
+            { "wet", &weather_effect::wet_player },
+            { "thunder", &weather_effect::thunder },
+            { "lightning", &weather_effect::lightning },
+            { "light_acid", &weather_effect::light_acid },
+            { "acid", &weather_effect::acid }
+        };
+        bool found = false;
+        for( const std::pair<const std::string, weather_effect_fn> &effect : all_weather_effects ) {
+            if( effect.first == pair.first ) {
+                weather_data.effects.emplace_back( effect.second, pair.second );
+                found = true;
+                break;
+            }
+        }
+        if( !found ) {
+            debugmsg( "Invalid weather effect %s", pair.first );
```

Comment:
```suggestion
            weather_effect.throw_error( "Invalid weather effect", "name" );
```
Don't even allow the game to continue loading with invalid data.